consensus/sync/sync_test.go

```diff
@@ -0,0 +1,171 @@
+	"github.com/NethermindEth/juno/consensus/starknet"
+	consensusSync "github.com/NethermindEth/juno/consensus/sync"
+	"github.com/NethermindEth/juno/consensus/tendermint"
+	"github.com/NethermindEth/juno/consensus/types"
+	"github.com/NethermindEth/juno/core"
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/db/pebble"
+	"github.com/NethermindEth/juno/p2p/sync"
+	"github.com/NethermindEth/juno/utils"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/mock/gomock"
+)
+
+var (
+	comittedHeight = -1
+	blockID        = uint64(9)
+)
+
+type (
+	listeners    = p2p.Listeners[starknet.Value, starknet.Hash, starknet.Address]
+	broadcasters = p2p.Broadcasters[starknet.Value, starknet.Hash, starknet.Address]
+	tendermintDB = db.TendermintDB[starknet.Value, starknet.Hash, starknet.Address]
+)
+
+func mockTimeoutFn(step types.Step, round types.Round) time.Duration {
+	return 10 * time.Second
+}
+
+func newTendermintDB(t *testing.T) tendermintDB {
+	t.Helper()
+	dbPath := t.TempDir()
+	pebbleDB, err := pebble.New(dbPath)
+	require.NoError(t, err)
+
+	return db.NewTendermintDB[starknet.Value, starknet.Hash, starknet.Address](pebbleDB, types.Height(0))
+}
+
+func newDB(t *testing.T) *mocks.MockTendermintDB[starknet.Value, starknet.Hash, starknet.Address] {
+	t.Helper()
+	ctrl := gomock.NewController(t)
+	// Ignore WAL for tests that use this
+	db := mocks.NewMockTendermintDB[starknet.Value, starknet.Hash, starknet.Address](ctrl)
+	db.EXPECT().GetWALEntries(gomock.Any()).AnyTimes()
+	db.EXPECT().SetWALEntry(gomock.Any()).AnyTimes()
+	db.EXPECT().Flush().AnyTimes()
+	db.EXPECT().DeleteWALEntries(gomock.Any()).AnyTimes()
+	return db
+}
+
+func TestSync(t *testing.T) {
+	ctrl := gomock.NewController(t)
+
+	ctx, cancel := context.WithCancel(t.Context())
+
+	nodeAddr := starknet.Address(felt.FromUint64(123))
+	logger := utils.NewNopZapLogger()
+	tendermintDB := newDB(t)
+	proposalStore := proposal.ProposalStore[starknet.Hash]{}
+	allNodes := newNodes(4) // Proposer set to 5
+	mockApp := mocks.NewMockApplication[starknet.Value, starknet.Hash](ctrl)
+	mockApp.EXPECT().Valid(gomock.Any()).AnyTimes().Return(true)
+	mockCommitListener := mocks.NewMockCommitListener[starknet.Value, starknet.Hash, starknet.Hash](ctrl)
+
+	mockCommitListener.EXPECT().
+		Commit(gomock.Any(), types.Height(0), gomock.Any()).
+		Do(func(_ any, newHeight types.Height, _ any) {
+			comittedHeight = int(newHeight)
+			cancel()
+		})
+
+	stateMachine := tendermint.New(tendermintDB, logger, nodeAddr, mockApp, allNodes, types.Height(0))
+
+	proposalCh := make(chan starknet.Proposal)
+	prevoteCh := make(chan starknet.Prevote)
+	precommitCh := make(chan starknet.Precommit)
+	p2p := newMockP2P(proposalCh, prevoteCh, precommitCh)
+
+	driver := driver.New(
+		utils.NewNopZapLogger(),
+		newTendermintDB(t),
+		stateMachine,
+		mockCommitListener,
+		p2p,
+		mockTimeoutFn,
+	)
+
+	go func() {
+		require.NoError(t, driver.Run(ctx))
```

Comment:
We don't need the driver here anymore, because the subject under test writes data to the channel, so it doesn't need to know about driver.